packages/core/mesh/network-manager/src/swarm/connection-limiter.ts

```diff
@@ -0,0 +1,71 @@
+//
+// Copyright 2023 DXOS.org
+//
+
+import { DeferredTask } from '@dxos/async';
+import { Context } from '@dxos/context';
+
+import { NetworkManager } from '../network-manager';
+import { ConnectionState } from './connection';
+
+export const MAX_INITIATING_CONNECTIONS = 15;
+
+export class ConnectionLimiter {
+  private readonly _ctx = new Context();
+
+  /**
+   * Queue of promises to resolve when initiating connections amount is below the limit.
+   */
+  private readonly _waitingPromises: { resolve: () => void }[] = [];
+
+  private _initiatingConnections = 0;
+
+  private readonly _updateInitiatingConnections = new DeferredTask(this._ctx, async () => {
```

Comment:
Instead of limiter gathering connection states from network manager, perhaps Peer would just notify connection limiter about connections opening and closing by calling methods on this class. This way ConnectionLimiter doesnt need to depend on the network manager